### PR TITLE
Add Exception configuration documentation

### PR DESCRIPTION
**Pull request by GregoireHebert**

### CODE REVIEW

File: core/exceptions.md [hunk 1]
```diff
@@ -0,0 +1,94 @@
+# Exceptions
```
Comment (dunglas) [inline core/exceptions.md]: I would rename this entry "Error handling" (`errors.md`).

File: core/exceptions.md [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+```YAML
+# app/config/config.yml
+
+ api_platform:
+
+    # ...
+
+    exception_to_status:
+        # By default there are two exceptions defined that you can customize.
```
Comment (dunglas) [inline core/exceptions.md]:
```yaml
# app/config/config.yml

services:
# Map exceptions to HTTP status codes using the `exception_to_status` configuration key 
exception_to_status:
    # The 2 following exceptions are handled by default
    Symfony\Component\Serializer\Exception\ExceptionInterface: 400                     # Use a raw status code (recommended)
    ApiPlatform\Core\Exception\InvalidArgumentException: 'HTTP_BAD_REQUEST' # Or with a constant of `Symfony\Component\HttpFoundation\Response`

    AppBundle\Exception\MyException: 401 # Custom exceptions can easily be handled
```

File: core/exceptions.md [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+        Foo\Bar\BazException: 401
+
+        # ...
+```
+
+## Custom Exception
+
+As in any php application, your exceptions have to extends the \Exception class or any of it's children.
+
+```PHP
```
Comment (dunglas) [inline core/exceptions.md]: php

File: core/exceptions.md [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+<?php
+
```
Comment (dunglas) [inline core/exceptions.md]: // src/AppBundle/Exception/ProductNotFoundException.php

File: core/exceptions.md [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+namespace AppBundle\Exception;
+
+/**
+ * Class ProductNotFoundException
```
Comment (dunglas) [inline core/exceptions.md]: I would remove this docblock.

File: core/exceptions.md [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+ */
+class ProductNotFoundException extends \Exception
```
Comment (dunglas) [inline core/exceptions.md]: `final`

File: core/exceptions.md [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+{
+
+    public function __construct($code = 0) {
```
Comment (dunglas) [inline core/exceptions.md]: I would remove this constructor.

File: core/exceptions.md [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+        parent::__construct("Product Not found", $code);
+    }
+}
+```
+
+```PHP
```
Comment (dunglas) [inline core/exceptions.md]: php

File: core/exceptions.md [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+<?php
+
```
Comment (dunglas) [inline core/exceptions.md]: // src/AppBundle/Manager/CartManager.php

File: core/exceptions.md [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+namespace AppBundle\Manager;
+
+use AppBundle\Entity\Product;
+use AppBundle\Exception\ProductNotFoundException;
+
+/**
+ * Class CartManager
```
Comment (dunglas) [inline core/exceptions.md]: I would remove this constructor.

File: core/exceptions.md [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+ */
+class CartManager
```
Comment (dunglas) [inline core/exceptions.md]: `final`

File: core/exceptions.md [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+{
+    const NOTFOUND_DOESNOTEXISTS = 51;
```
Comment (dunglas) [inline core/exceptions.md]: I would strip the `NOTFOUND` prefix and a visibility to all those constants.

File: core/exceptions.md [hunk 13]
```diff
@@ -0,0 +1,94 @@
...
+    const NOTFOUND_OUTOFSTOCK = 53;
+
+    /**
+     * @param integer $id
+     *
+     * @throws ProductNotFoundException
+     *
+     * @return Product $product
+     */
+    public function addProductToCart($id)
```
Comment (dunglas) [inline core/exceptions.md]: `public function addProductToCart(int $id)`

File: core/exceptions.md [hunk 14]
```diff
@@ -0,0 +1,94 @@
...
+ * Class CartManager
+ */
+class CartManager
+{
+    const NOTFOUND_DOESNOTEXISTS = 51;
+    const NOTFOUND_DEACTIVATED = 52;
+    const NOTFOUND_OUTOFSTOCK = 53;
+
+    /**
+     * @param integer $id
```
Comment (dunglas) [inline core/exceptions.md]: int

File: core/exceptions.md [hunk 15]
```diff
@@ -0,0 +1,94 @@
...
+        $product = /*...*/;
+
+        if (!$product->getVirtualStock()) {
+            // Using internal codes for a better understanding of what's going on
+            throw new ProductNotFoundException(self::NOTFOUND_OUTOFSTOCK);
+        }
+
+        /*...*/
+
+        return $cart;
```
Comment (dunglas) [inline core/exceptions.md]: I would return nothing (CQRS like)

File: core/exceptions.md [hunk 16]
```diff
@@ -0,0 +1,94 @@
...
+    public function addProductToCart($id)
+    {
+        $product = /*...*/;
+
+        if (!$product->getVirtualStock()) {
+            // Using internal codes for a better understanding of what's going on
+            throw new ProductNotFoundException(self::NOTFOUND_OUTOFSTOCK);
+        }
+
+        /*...*/
```
Comment (dunglas) [inline core/exceptions.md]: Can be removed

**Comment by dunglas** [comment]
I would merge both sections (the last section should be an example illustrating the config). I propose to start with the tiny exception example (it can simplified IMO, and it should be more integrated with API Platform, for instance used as an event listener).

**Comment by dunglas** [comment]
By the way it would be nice to show that API Platform automatically handle validation errors.

File: core/exceptions.md [hunk 17]
```diff
@@ -0,0 +1,94 @@
...
+
+        return $cart;
+    }
+}
+```
+
+It doesn't have to be an HttpException, any type of Exception can be thrown. The best part is that API Platform already takes care of how the error is handled and returned. If you configured your API to respond in a JsonLD format, your error will be returned in JsonLD format as well.
```
Comment (dunglas) [inline core/exceptions.md]:
```
The exception doesn't have to be a Symfony's `HttpException`. Any type of `Exception` can be thrown. The best part is that API Platform already takes care of how the error is handled and returned. For instance the API is configured to respond in JSON-LD, the error will be returned in this format as well.
```

**Comment by GregoireHebert** [comment]
I'll do the example for validation errors and transform the error as an event listener tomorrow.

**Comment by GregoireHebert** [comment]
Ok I made some changes, but I not really convinced that what you had in mind.
Let me know :)

**Pull request closed without merging**